### PR TITLE
[PERF] Fuse forward + backward + optimizer into a single graph

### PR DESCRIPTION
**Pull request by abogdanovicTT**
### Issue

N/A

### Problem description

The training step runs the forward+backward and the optimizer as two separate graphs. Two things force extra compilations / executions:

1. After `loss_.backward()` the loss and all grads are synced with `_xla_sync_multi` + `_clear_pending_irs`, which cuts the graph between backward and the optimizer step. Forward+backward and the optimizer update therefore compile and execute as separate graphs every step.
2. `torch.optim.AdamW` lazily allocates its `step` / `exp_avg` / `exp_avg_sq` state on the first `step()`. This changes the optimizer graph's input signature after step 0 and forces a second compilation of the step.

This PR fuses forward + backward + optimizer into a single graph and pre-materializes the AdamW state so the fused step compiles only once.

### What's Changed

- Fuse the whole step into one graph in `train.py`: drop the intermediate `_xla_sync_multi` / `_clear_pending_irs` on loss+grads so the forward/backward IR stays pending and joins the optimizer update, executed by the single sync inside `optimizer_step`.
- Add `materialize_adamw_state` to `tools/workaround_utils.py`: pre-initializes AdamW `step`/`exp_avg`/`exp_avg_sq` to zero (a numerical no-op, since Adam's own lazy init also starts at zero) so the fused step graph keeps a stable input signature and compiles once.
- Accumulate a detached `step_loss` lazily and read it (`.item()`) only after `optimizer_step`, so logging the loss no longer cuts the graph.

### CODE REVIEW

File: blacksmith/experiments/torch/llama/xla/lora/single_chip/llama_3_1_8b_sst2.yaml [hunk 1]
```diff
@@ -32,8 +32,8 @@ wandb_tags: ["test"]
 wandb_watch_mode: "all"
 wandb_log_freq: 1000
 model_to_wandb: False
-steps_freq: 10
-val_steps_freq: 10
```
Comment (mmilosevicTT) [inline blacksmith/experiments/torch/llama/xla/lora/single_chip/llama_3_1_8b_sst2.yaml:36]: This does not seem related to the PR.

File: blacksmith/experiments/torch/llama/xla/train.py [hunk 2]
```diff
@@ -192,20 +201,20 @@ def train(
...
-                    # Optimizer will do unnecessary recompute if we don't clear the pending IRs after syncing the loss and grads.
-                    torch_xla._XLAC._clear_pending_irs(torch_xla._XLAC._xla_get_default_device())
-
-                running_loss += loss_.item()
+                # Keep the whole step (forward + backward + optimizer) in a single graph.
+                # No sync is issued here on purpose: the forward and backward IR stays pending
+                # and lazily joins the optimizer update, so everything is compiled and executed
+                # together by the one sync inside `optimizer_step`. The detached micro-batch loss
+                # is accumulated lazily so it can be read afterwards without cutting the graph.
+                step_loss = loss_.detach() if step_loss is None else step_loss + loss_.detach()
```
Comment (mmilosevicTT) [inline blacksmith/experiments/torch/llama/xla/train.py]: So this means we are extending the graph to grad_accumulation_step loops + optimizer? Will this pose problems with memory?

File: blacksmith/experiments/torch/llama/xla/train.py [hunk 3]
```diff
@@ -113,6 +117,10 @@ def train(
     trainable_params = [p for p in model.parameters() if p.requires_grad]
     optimizer = torch.optim.AdamW(trainable_params, capturable=config.use_tt, lr=config.learning_rate)
 
+    # Pre-create optimizer state so the fused step graph compiles only once.
+    if config.use_tt and not config.resume_from_checkpoint:
+        materialize_adamw_state(optimizer)
+
     # Load checkpoint if needed.
     if config.resume_from_checkpoint:
         checkpoint_manager.load_checkpoint(model, optimizer)
```
Comment (mmilosevicTT) [inline blacksmith/experiments/torch/llama/xla/train.py:141]: Did you try loading from checkpoint? Will this always stay one graph? i.e does load_checkpoint not introduce some operations like it was with fresh optimizer?

**Review by mmilosevicTT — APPROVED** [review]
LGTM

Left few comments

**Review by agobeljicTT — APPROVED** [review]

File: blacksmith/experiments/torch/llama/xla/train.py (hunk 2 above)
Comment (agobeljicTT) [reply blacksmith/experiments/torch/llama/xla/train.py]: fixed it in the last commit, now we have one fwd+bwd and one fwd+bwd+opt

File: blacksmith/experiments/torch/llama/xla/train.py (hunk 3 above)
Comment (agobeljicTT) [reply blacksmith/experiments/torch/llama/xla/train.py:141]: i'll try this Monday, make myself a reminder

Comment (agobeljicTT) [reply blacksmith/experiments/torch/llama/xla/train.py:141]: there was an issue, resolved it by putting optimizer to device and syncing
